8260931: Implement JEP 382: New macOS Rendering Pipeline

aghaisas opened the pull request:
**Description :**
This is the implementation of [JEP 382 : New macOS Rendering Pipeline](https://bugs.openjdk.java.net/browse/JDK-8238361)
It implements a Java 2D internal rendering pipeline for macOS using the Apple Metal API.
The entire work on this was done under [OpenJDK Project - Lanai](http://openjdk.java.net/projects/lanai/)

We iterated through several Early Access (EA) builds and have reached a stage where it is ready to be integrated to openjdk/jdk. The latest EA build is available at - https://jdk.java.net/lanai/

A new option -Dsun.java2d.metal=true | True needs to be used to use this pipeline.

**Testing :**
This implementation has been tested with the tests present at - [Test Plan for JEP 382: New macOS Rendering Pipeline](https://bugs.openjdk.java.net/browse/JDK-8251396)

**Note to reviewers :**
1) Default rendering pipeline on macOS has not been changed by this PR. OpenGL still stays as the default rendering pipeline and Metal rendering pipeline is optional to choose.

2) To apply and test this PR - 
To enable the metal pipeline you must specify on command line -Dsun.java2d.metal=true (No message will be printed in this case) or -Dsun.java2d.metal=True (A message indicating Metal rendering pipeline is enabled gets printed)

3) Review comments (including some preliminary informal review comments) are tracked with JBS issues - https://bugs.openjdk.java.net/issues/?filter=40598

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260931](https://bugs.openjdk.java.net/browse/JDK-8260931): Implement JEP 382: New macOS Rendering Pipeline


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 0b8d96b2cef5327e0d97fbdf8121748d891f306f
 * [Alexey Ushakov](https://openjdk.java.net/census#avu) (@avu - no project role) ⚠️ Review applies to 3980ecbd2a9769167c33b07e4199d3a53e6e0c43
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - Author)
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to 7b0b0dc42d219dcda270492667cd0f37b9e2a70a
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**)
 * [Jayathirth D V](https://openjdk.java.net/census#jdv) (@jayathirthrao - **Reviewer**)
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)


### Contributors
 * Jayathirth D V `<jdv@openjdk.org>`
 * Alexey Ushakov `<avu@openjdk.org>`
 * Artem Bochkarev `<abochkarev@openjdk.org>`
 * Prasanta Sadhukhan `<psadhukhan@openjdk.org>`
 * Denis Konoplev `<dkonoplev@openjdk.org>`
 * Phil Race `<prr@openjdk.org>`
 * Kevin Rushforth `<kcr@openjdk.org>`
 * Magnus Ihse Bursie `<ihse@openjdk.org>`
 * Ajit Ghaisas `<aghaisas@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2403/head:pull/2403`
`$ git checkout pull/2403`

aghaisas commented:
/reviewers 3

aghaisas commented:
/contributor add jdv

aghaisas commented:
/contributor add avu

aghaisas commented:
/contributor add psadhukhan

aghaisas commented:
/contributor add dkonoplev

aghaisas commented:
/contributor add prr

aghaisas commented:
/contributor add kcr

aghaisas commented:
/contributor add ihse

aghaisas commented:
/csr

aghaisas commented:
/contributor add aghaisas

magicus reviewed (approved):
Build changes look good. No opinion on the source changes.

avu commented on src/java.desktop/share/native/libawt/java2d/SurfaceData.c:
```diff
@@ -235,6 +235,7 @@ JNIEXPORT SurfaceDataOps * JNICALL
 SurfaceData_InitOps(JNIEnv *env, jobject sData, int opsSize)
 {
     SurfaceDataOps *ops = malloc(opsSize);
+
```
The only whitespace added. Should be removed

avu reviewed (changes requested):
Looks good except for one whitespace problem (src/java.desktop/share/native/libawt/java2d/SurfaceData.c)

kevinrushforth commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/EncoderManager.h:
```diff
@@ -0,0 +1,95 @@
+ *  for example destination, stencil, or any other property of MTLRenderPassDescriptor)
+ *  2. Updates 'mutable' properties encoder: pipelineState (with corresponding buffers), clip, transform, e.t.c. To avoid
+ *  unnecessary calls of [encoder setXXX] this manager compares requested state with cached one.
+ * */
```
Minor: `* */` --> `*/`

kevinrushforth commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/EncoderManager.m:
```diff
@@ -0,0 +1,437 @@
+- (void) updateClip:(id<MTLRenderCommandEncoder>)encoder clip:(MTLClip *)clip forceUpdate:(jboolean)forceUpdate
+{
+    if (clip.stencilMaskGenerationInProgress == JNI_TRUE) {
+        // don't set setScissorOrStencil when generateion in progress
```
Minor: typo in comment, should be `generation`.

kevinrushforth reviewed (commented):
Here is my initial set of mostly minor comments and a couple questions.

kevinrushforth reviewed (commented):
I left a few additional comments. Overall this looks good to me.

kevinrushforth commented on test/jdk/performance/client/RenderPerfTest/Makefile:
```diff
@@ -0,0 +1,86 @@
+#
+#   - Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#
```
As mentioned in the preliminary review, this file and `build.xml` have a BSD copyright. I think that should be GPLv2 (without classpath exception since this is a test).

mrserb replied:
Probably place it near J2Dbench?

kevinrushforth commented on test/jdk/performance/client/RenderPerfTest/src/renderperf/RenderPerfTest.java:
```diff
@@ -0,0 +1,955 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
```
Since this a test, this should not have the Classpath exception.

kevinrushforth reviewed (commented):
The file in `RenderPerfTest` should have a GPLv2 license header (no Classpath). I filed [JDK-8261273](https://bugs.openjdk.java.net/browse/JDK-8261273) and also highlighted a couple examples below.

aghaisas commented on make/modules/java.desktop/lib/Awt2dLibraries.gmk:
```diff
@@ -884,6 +888,28 @@ endif
+  SHADERS_SRC := $(TOPDIR)/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/shaders.metal
+  SHADERS_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/native/java.desktop/libosxui
+  SHADERS_AIR := $(SHADERS_SUPPORT_DIR)/shaders.air
+  SHADERS_LIB := $(INSTALL_LIBRARIES_HERE)/shaders.metallib
```
I think, a generic name is OK as the path of shader file already has both awt (libawt_lwawt) and java2d in it.

kevinrushforth replied:
In the source tree, yes, but not in the jdk image where it ends up in `$JAVA_HOME/lib/shaders.metallib`. I don't have a problem with this, as long as it is a deliberate decision.

aghaisas replied:
OK. I get your point. Keeping the name unchanged for now as there won't be another .metallib in JDK. The name can be changed in the future if need arises.

gerard-ziemski commented:
General comment - I am not sure I like the `MTL` prefix acronym in names (ex. `sun.java2d.metal.MTLVolatileSurfaceManager`).

I think you tried to match the `CGL`, but that is a real acronym that stands for "Core Graphics Layer" (I think).

`MTL` on the other hand is no acronym. I can see `ML` for "Metal Layer" I suppose, but also just `Metal` would work just fine.

gerard-ziemski commented:
I'm in the process of reviewing this feature, but there is lots of code to go through - please wait for my review.

kevinrushforth commented:
> General comment - I am not sure I like the `MTL` prefix acronym in names (ex. `sun.java2d.metal.MTLVolatileSurfaceManager`).
> 
> I think you tried to match the `CGL`, but that is a real acronym that stands for "Core Graphics Layer" (I think).
> 
> `MTL` on the other hand is no acronym. I can see `ML` for "Metal Layer" I suppose, but also just `Metal` would work just fine.

`MTL` is the abbreviation that Apple uses for Metal in all of their APIs. The only potential issue I might see with this prefix is in the native code where there could be name collisions between Java2D's names and Apple's names. Since we haven't run into such a collision, I don't think this needs to change, and wouldn't necessarily affect the Java class names anyway. If we were to consider it, `METAL` seems better than `ML` (which is too short to be descriptive and might suggest machine learning).

gerard-ziemski commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLRenderQueue.java:
```diff
@@ -0,0 +1,254 @@
+     * graphics config info pointer on the single queue flushing thread.
+     */
+    public static void disposeGraphicsConfig(long pConfigInfo) {
+        MTLRenderQueue rq = getInstance();
+        rq.lock();
```
Is it allowed to have multiple `MTLRenderQueue` instances?

If not, then I see this pattern everywhere:

```java
        MTLRenderQueue rq = getInstance();
        rq.lock();
        {
          ...
        }
        rq.unlock();
```
why not just do:

```java
        MTLRenderQueue.lock();
        {
          ...
        }
        MTLRenderQueue.unlock();
```

and have `MTLRenderQueue.lock()` and `MTLRenderQueue.unlock()` implement getting the actual lock instance and locking/unlocking it?

jayathirthrao replied:
Thanks for the recommendation. This is a common behavior among different pipelines :  D3D, OpenGL and Metal. Mentioned lock/unlock should be implemented in parent RenderQueue class after refactoring and it will hit other pipelines. Its better to not touch other pipelines as part of this JEP.

gerard-ziemski reviewed (commented):
> > General comment - I am not sure I like the `MTL` prefix acronym in names (ex. `sun.java2d.metal.MTLVolatileSurfaceManager`).
> > I think you tried to match the `CGL`, but that is a real acronym that stands for "Core Graphics Layer" (I think).
> > `MTL` on the other hand is no acronym. I can see `ML` for "Metal Layer" I suppose, but also just `Metal` would work just fine.
> 
> `MTL` is the abbreviation that Apple uses for Metal in all of their APIs. The only potential issue I might see with this prefix is in the native code where there could be name collisions between Java2D's names and Apple's names. Since we haven't run into such a collision, I don't think this needs to change, and wouldn't necessarily affect the Java class names anyway. If we were to consider it, `METAL` seems better than `ML` (which is too short to be descriptive and might suggest machine learning).

If Apple itself uses `MTL` then we are good.

gerard-ziemski commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLSurfaceData.java:
```diff
@@ -0,0 +1,649 @@
+import static sun.java2d.pipe.hw.ContextCapabilities.CAPS_PS30;
+
+
+public abstract class MTLSurfaceData extends SurfaceData
+        implements AccelSurface {
```
`MTLSurfaceData` and `OGLSurfaceData` seem to share lots of the same code, isn't there a way to refactor the common code out?

There are other files that structually look identical, except for the names of classes they use, ex `MTLMaskBlit.java` and `OGLMaskBlit.java`.

aghaisas replied:
Although it is a good suggestion, I do not recommend this refactoring - 
- We do not wish to destabilize proven OpenGL pipeline by refactoring - additional testing on OpenGL pipeline also would be needed.
- OpenGL pipeline on macOS might be removed once Apple removes OpenGL support from macOS - this will render this refactoring pointless.

aghaisas commented on src/java.desktop/macosx/classes/sun/awt/CGraphicsDevice.java:
```diff
@@ -63,7 +68,57 @@
+        if (!metalPipelineEnabled && !oglPipelineEnabled) {
+            // This indicates fallback to other rendering pipeline also failed.
+            // Should never reach here
+            throw new InternalError("Error - unable to initialize any rendering pipeline.");
```
OpenGL was the sole rendering pipeline on macOS. Now we are adding Metal rendering pipeline. There is no software based fall back renderer.

aghaisas commented on src/java.desktop/macosx/classes/sun/java2d/MacOSFlags.java:
```diff
@@ -0,0 +1,165 @@
+        String propString = System.getProperty(p);
+        PropertyState returnVal = defaultVal;
+        if (propString != null) {
+            if (propString.equals("true") ||
+                propString.equals("t") ||
+                propString.equals("True") ||
+                propString.equals("T") ||
+                propString.equals("")) // having the prop name alone
+            {                          // is equivalent to true
+                returnVal = PropertyState.ENABLED;
+            } else if (propString.equals("false") ||
+                       propString.equals("f") ||
+                       propString.equals("False") ||
+                       propString.equals("F"))
```
No.
0 and 1 are not supported for existing VM options for rendering pipelines. We won't be adding it for metal pipeline either.

aghaisas commented on src/java.desktop/macosx/classes/sun/java2d/MacOSFlags.java:
```diff
@@ -0,0 +1,165 @@
+                    // Handle invalid combinations to use the default rendering pipeline
+                    // Current default rendering pipeline is OpenGL
+                    // (The default can be changed to Metal in future just by toggling two states in this if condition block)
+                    if ((oglState == PropertyState.UNSPECIFIED && metalState == PropertyState.UNSPECIFIED) ||
+                        (oglState == PropertyState.DISABLED && metalState == PropertyState.DISABLED) ||
+                        (oglState == PropertyState.ENABLED && metalState == PropertyState.ENABLED)) {
+                        oglState = PropertyState.ENABLED; // Enable default pipeline
+                        metalState = PropertyState.DISABLED; // Disable non-default pipeline
+                    }
```
OpenGL was the sole rendering pipeline on macOS. Now we are adding Metal rendering pipeline. There is no software based fall back renderer.

If user requests both - GL and Metal - as false - we shall use 'the default' pipeline.
Current default render pipeline on macOS is OpenGL.
Changing default render pipeline on macOS is not in the scope of this JEP.

prsadhuk commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLBlitLoops.java:
```diff
@@ -0,0 +1,959 @@
+//                 new MTLSwToSurfaceBlit(SurfaceType.ByteGray,
+//                         MTLSurfaceData.PF_BYTE_GRAY),
+//                 new MTLSwToSurfaceBlit(SurfaceType.UshortGray,
+//                         MTLSurfaceData.PF_USHORT_GRAY),
```
Probably we could remove this commented lines

gerard-ziemski commented:
I tried to code review the native implementation file, but Metal APIs is brand new to me and it's been a long while since I worked with graphics API, so I can't be of much help really.

The code I've looked at looked clean and nothing caught my eye. But it's a partial code review at best.

Good job!

gerard-ziemski commented:
A small thing caught my eye when I was comparing OpenGL and Metal pipelines running J2Demo.jar

It looks to me like OpenGL rendering pipeline uses brighter colors. I prefer the more subdued colors that Metal is rendering, but not sure if this is something you want to investigate further.

I tried to capture what I see below:

OpenGL:

<img width="1128" alt="OpenGL" src="https://user-images.githubusercontent.com/63425797/107575744-dc0ca180-6bb5-11eb-9eb3-5cff415eb8a3.png">

Metal (the green color has a yellowish tint):

<img width="1125" alt="Metal" src="https://user-images.githubusercontent.com/63425797/107575716-cf884900-6bb5-11eb-8ae8-14212ec94e87.png">

Performance wise I did not see much difference either way.

gerard-ziemski commented:
According to Xcode Instruments leak profile, there are 2 minor memory leaks in the Metal rendering pipeline:

`#1 Malloc 80 Bytes	1	0x7fde0d4247b0	80 Bytes	libjava.dylib	getStringUTF8`
```c
   0 libsystem_malloc.dylib malloc_zone_malloc
   1 libsystem_malloc.dylib malloc
   2 libjava.dylib getStringUTF8 /Volumes/Work/review/2403/jdk/src/java.base/share/native/libjava/jni_util.c:888
   3 libjava.dylib JNU_GetStringPlatformChars /Volumes/Work/review/2403/jdk/src/java.base/share/native/libjava/jni_util.c:917
   4 libawt_lwawt.dylib Java_sun_java2d_metal_MTLGraphicsConfig_getMTLConfigInfo /Volumes/Work/review/2403/jdk/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLGraphicsConfig.m:151
   5  0x11ab08d48
   6  0x11ab0250d
```

`#2 Malloc 80 Bytes	1	0x7fde0d4325a0	80 Bytes	libjava.dylib	getStringUTF8`
```c
   0 libsystem_malloc.dylib malloc_zone_malloc
   1 libsystem_malloc.dylib malloc
   2 libjava.dylib getStringUTF8 /Volumes/Work/review/2403/jdk/src/java.base/share/native/libjava/jni_util.c:888
   3 libjava.dylib JNU_GetStringPlatformChars /Volumes/Work/review/2403/jdk/src/java.base/share/native/libjava/jni_util.c:917
   4 libawt_lwawt.dylib Java_sun_java2d_metal_MTLGraphicsConfig_tryLoadMetalLibrary /Volumes/Work/review/2403/jdk/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLGraphicsConfig.m:120
   5  0x11ab08d48
   6  0x11ab024c8
```

Those can be handled as a followup issues though if you like, it's only 160 bytes total.

aghaisas commented:
> I tried to code review the native implementation files, but Metal APIs is brand new to me and it's been a long while since I worked with graphics API, so I can't be of much help really.
> 
> The code I've looked at looked clean and nothing caught my eye. But it's a partial code review at best.
> 
> Good job!

Every bit helps. Thanks for your review effort!

jayathirthrao commented:
> A small thing caught my eye when I was comparing OpenGL and Metal pipelines running J2Demo.jar
> 
> It looks to me like OpenGL rendering pipeline uses brighter colors. I prefer the more subdued colors that Metal is rendering, but not sure if this is something you want to investigate further.
> 
> I tried to capture what I see below:
> 
> OpenGL:
> 
> <img alt="OpenGL" width="1128" src="https://user-images.githubusercontent.com/63425797/107575744-dc0ca180-6bb5-11eb-9eb3-5cff415eb8a3.png">
> 
> Metal (the green color has a yellowish tint):
> 
> <img alt="Metal" width="1125" src="https://user-images.githubusercontent.com/63425797/107575716-cf884900-6bb5-11eb-8ae8-14212ec94e87.png">
> 
> Performance wise I did not see much difference either way.

@gerard-ziemski Thanks for verifying the behavior.
Text present in Memory monitor/Performance in J2DDemo is not drawn using Metal rendering pipeline(drawGlyphList or any other rendering opcodes, I have verified it again in latest code). This is drawn using software renderloops and only thing happens from Metal pipeline is this content is used in blit opcode. Slight difference in color that we are seeing is related to how Metal internally handles blending of colors.

avu commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTexurePool.m:
```diff
@@ -0,0 +1,443 @@
+#import "MTLTexturePool.h"
+#import "Trace.h"
+
+#define SCREEN_MEMORY_SIZE_4K (4096*2160*4) //~33,7 mb
```
This parameter manages our caching strategy for the temporary texture pool. Huge texture allocations 4K/2 size will cause texture pool drain. We need some more profiling to just these parameters. And I think we need to use Metal API (https://developer.apple.com/documentation/metal/mtldevice/2369280-recommendedmaxworkingsetsize?language=objc) in the future to adjust the amount of memory that we should use.

gerard-ziemski commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLLayer.m:
```diff
@@ -0,0 +1,272 @@
+        self.nextDrawableCount++;
+        id <MTLBlitCommandEncoder> blitEncoder = [commandBuf blitCommandEncoder];
+
+        [blitEncoder
+                copyFromTexture:self.buffer sourceSlice:0 sourceLevel:0
+                sourceOrigin:MTLOriginMake((jint)(self.leftInset*self.contentsScale), (jint)(self.topInset*self.contentsScale), 0)
+                sourceSize:MTLSizeMake(self.buffer.width, self.buffer.height, 1)
+                toTexture:mtlDrawable.texture destinationSlice:0 destinationLevel:0 destinationOrigin:MTLOriginMake(0, 0, 0)];
+        [blitEncoder endEncoding];
```
There is an issue with this code. Running Java2D.jar in Xcode asserts here with this message:

```c
2021-02-11 14:11:45.710457-0600 java[49971:9486360] Metal API Validation Enabled
2021-02-11 14:11:46.038720-0600 system_profiler[49975:9486885] Metal API Validation Enabled
-[MTLDebugBlitCommandEncoder internalValidateCopyFromTexture:sourceSlice:sourceLevel:sourceOrigin:sourceSize:toTexture:destinationSlice:destinationLevel:destinationOrigin:options:]:474: failed assertion `(sourceOrigin.y + sourceSize.height)(444) must be <= height(400).'
(lldb) 
```

and forcing the execution past it results in the java process crashing.

A solution would be to clip the src size like this:

```c
        NSUInteger src_x = self.leftInset*self.contentsScale;
        NSUInteger src_y = self.topInset*self.contentsScale;
        NSUInteger src_w = self.buffer.width-src_x;
        NSUInteger src_h = self.buffer.height-src_y;
        
        [blitEncoder
                copyFromTexture:self.buffer sourceSlice:0 sourceLevel:0
                sourceOrigin:MTLOriginMake((jint)src_x, (jint)src_y, 0)
                sourceSize:MTLSizeMake(src_w, src_h, 1)
                toTexture:mtlDrawable.texture destinationSlice:0 destinationLevel:0 destinationOrigin:MTLOriginMake(0, 0, 0)];
        [blitEncoder endEncoding];
```

gerard-ziemski replied:
I guess you will only see this if `Metal API Validation Enabled`

gerard-ziemski replied:
Which actually begs a question of whether we tested with `Metal API Validation Enabled` ?

prrace replied:
I submitted https://bugs.openjdk.java.net/browse/JDK-8261620 for this bug.
Could be that there are more such issues but since this crash occurs on start up I can't say what else there might be.

aghaisas replied:
Thanks @gerard-ziemski for bringing out this important observation. It clearly indicates that there are some overflows which are undetected.

We have tested without Metal API validation enabled till now.
 
Now, I have fixed the JDK-8261620 bug that reported the crash on app launch with Metal API validation enabled.
More testing with “Metal API validation enabled” to follow..

gerard-ziemski commented:
I took a look at https://bugs.openjdk.java.net/browse/JDK-8261408 and noticed a startup time regression with the Metal rendering pipeline, so I dug a bit and here is what I found using Xcode startup profiler:

Here is the OpenGL pipeline:
<img width="1904" alt="OpenGL" src="https://user-images.githubusercontent.com/63425797/107981859-8efb4780-6f88-11eb-849a-385c29f3ff6f.png">


Here is the Metal pipeline:
<img width="1904" alt="Metal" src="https://user-images.githubusercontent.com/63425797/107981882-9ae70980-6f88-11eb-9b73-d1bd19dfa07f.png">

If I read the results correctly, there is a weird 330ms time gap in the case of the Metal pipeline where nothing is done and it looks like the culprit is `Java_sun_java2d_metal_MTLGraphicsConfig_isMetalFrameworkAvailable`

The OpenGL pipeline gets blocked only for 83ms on `Java_sun_java2d_opengl_CGLGraphicsConfig_getCGLConfigInfo`

I would advice that we take a closer look at why 'Java_sun_java2d_metal_MTLGraphicsConfig_isMetalFrameworkAvailable' takes so long, and optimize it, or cache the results, so the net VM instance can skip it if needed (assuming the call doesn't end up initializing the native Metal or something like that), still worth taking a look.

I would also recommend that you adopt Xcode and its Instruments profiler tools for future work. Please let me know if you need help in this area.

I only scratched the surface here and I think that there is plenty of profiling that can be done to investigate the startup time regression further.

aghaisas commented:
Thanks @gerard-ziemski for taking a detailed look at this.
We do have an open bug to address this. Please refer [JDK-8259825](https://bugs.openjdk.java.net/browse/JDK-8259825).

prrace commented:
> Thanks @gerard-ziemski for taking a detailed look at this.
> We do have an open bug to address this. Please refer [JDK-8259825](https://bugs.openjdk.java.net/browse/JDK-8259825).

Hi Gerard, expecting a process and parsing its output is definitely not ideal and that's why there's this open bug.
One thing that is under consideration is to equate >= 10.14  with Metal is available since as of 10.14 macOS won't install if a system does not support metal. We have no compelling reason to support metal on earlier releases anyway .. those are either already unsupported or barely supported and OGL will always be available there.

gerard-ziemski commented:
> > Thanks @gerard-ziemski for taking a detailed look at this.
> > We do have an open bug to address this. Please refer [JDK-8259825](https://bugs.openjdk.java.net/browse/JDK-8259825).
> 
> Hi Gerard, expecting a process and parsing its output is definitely not ideal and that's why there's this open bug.
> One thing that is under consideration is to equate >= 10.14 with Metal is available since as of 10.14 macOS won't install if a system does not support metal. We have no compelling reason to support metal on earlier releases anyway .. those are either already unsupported or barely supported and OGL will always be available there.

I did not know that there already was a bug covering this issue.

Your idea seems reasonable.

More than just focusing on this immediate issue, however, I was hoping to raise the point of us starting adopting profiling tools to analyze our code (memory utilization, leaks, cpu/gpu profiling etc). A new feature that brings 10% benefit, but uses 50% more resources for example would probably not be a good investment. And to figure that we need to relay on some good tools and Xcode provides some.

prrace commented:
> > > Thanks @gerard-ziemski for taking a detailed look at this.
> > > We do have an open bug to address this. Please refer [JDK-8259825](https://bugs.openjdk.java.net/browse/JDK-8259825).
> > 
> > 
> > Hi Gerard, expecting a process and parsing its output is definitely not ideal and that's why there's this open bug.
> > One thing that is under consideration is to equate >= 10.14 with Metal is available since as of 10.14 macOS won't install if a system does not support metal. We have no compelling reason to support metal on earlier releases anyway .. those are either already unsupported or barely supported and OGL will always be available there.
> 
> I did not know that there already was a bug covering this issue.
> 
> Your idea seems reasonable.
> 
> More than just focusing on this immediate issue, however, I was hoping to raise the point of us starting adopting profiling tools to analyze our code (memory utilization, leaks, cpu/gpu profiling etc). A new feature that brings 10% benefit, but uses 50% more resources for example would probably not be a good investment. And to figure that we need to relay on some good tools and Xcode provides some.

There were actually tasks to do profiling of the memory footprint and look for leaks. We did not think it possible to be able to assert "Metal must use less memory than OpenGL" or dig into tiny differences but it was intended to find the big issues. See https://bugs.openjdk.java.net/browse/JDK-8237856

mrserb commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLSurfaceDataBase.h:
```diff
@@ -0,0 +1,150 @@
+#define MTLSD_WINDOW          sun_java2d_pipe_hw_AccelSurface_WINDOW
+#define MTLSD_TEXTURE         sun_java2d_pipe_hw_AccelSurface_TEXTURE
+#define MTLSD_FLIP_BACKBUFFER sun_java2d_pipe_hw_AccelSurface_FLIP_BACKBUFFER
+#define MTLSD_RT_TEXTURE        sun_java2d_pipe_hw_AccelSurface_RT_TEXTURE
```
Only two of them are supported? MTLSD_TEXTURE and MTLSD_RT_TEXTURE?

avu replied:
Also, MTLSD_WINDOW is supported

mrserb replied:
How it is used? Can we really draw to the window directly? It looks like it copied from the OGL where it is unused since CLayer integration. Something is inconsistent here.

avu replied:
Here is a follow-up issue JDK-8263463

avu commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLContext.java:
```diff
@@ -0,0 +1,154 @@
+     * situations where we may not otherwise have a current context (e.g.
+     * when disposing a texture-based surface).
+     */
+    public static void setScratchSurface(long pConfigInfo) {
```
In fact, we don't have any scratch surfaces. SET_SCRATCH_SURFACE effectively sets the new context. For better readability, we should add the new op SET_CONTEXT in BufferedOpCodes

prrace commented on test/jdk/performance/client/RenderPerfTest/build.xml:
```diff
@@ -0,0 +1,86 @@
+    <mkdir dir="${dist}"/>
+
+    <!-- Put everything in ${build} into the J2DBench.jar file -->
+    <jar jarfile="${dist}/RenderPerfTest.jar" basedir="${build}">
```
This was presumably borrowed from J2DBench so this comment should be fixed !

avu replied:
JDK-8263325

prrace commented:
Regarding this  comment from @mrserb 
> Probably place it near J2Dbench?

I can't find the context but if it is suggesting moving RenderPerfTest to be co-located with J2Bench let's NOT do that.
It is much  more likely that J2DBench will be moved out of demo into test ...

mrserb commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLBlitLoops.m:
```diff
@@ -0,0 +1,816 @@
+            width = srcInfo.bounds.x2 - srcInfo.bounds.x1;
+            height = srcInfo.bounds.y2 - srcInfo.bounds.y1;
+
+            pDst = PtrAddBytes(pDst, dstx * dstInfo.pixelStride);
```
I think this at least should be investigated before integration.

mrserb commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLGraphicsConfig.m:
```diff
@@ -0,0 +1,260 @@
+
+
+    NSRect contentRect = NSMakeRect(0, 0, 64, 64);
+    NSWindow *window =
```
Some remnant from the scratch surface initialization?

avu commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLSurfaceData.m:
```diff
@@ -0,0 +1,396 @@
+ * successful; JNI_FALSE otherwise.
+ */
+jboolean
+MTLSD_InitMTLWindow(JNIEnv *env, BMTLSDOps *bmtlsdo)
```
MTLContext setSurfacesEnv

avu reviewed (changes requested):
Looks good, but a couple of things should be fixed (JDK-8263325, JDK-8263324)

aghaisas commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLGraphicsConfig.m:
```diff
@@ -0,0 +1,267 @@
+Java_sun_java2d_metal_MTLGraphicsConfig_isMetalFrameworkAvailable
+    (JNIEnv *env, jclass mtlgc)
+{
+    FILE *f = popen("/usr/sbin/system_profiler SPDisplaysDataType", "r");
```
Fixed with JDK-8259825.

aghaisas commented:
We have addressed the review comments (except for a few minor follow-on issues)
This version of the PR is the one which we are planning to integrate into the JDK. Please review and approve.

kevinrushforth reviewed (approved):
Latest changes look good. I confirmed that this PR has all of the content from the lanai repo.

mrserb commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLGraphicsConfig.java:
```diff
@@ -0,0 +1,401 @@
+        MTLRenderQueue rq = MTLRenderQueue.getInstance();
+        rq.lock();
+        try {
+            // getMTLConfigInfo() creates and destroys temporary
```
Does it really create and destroy temporary surfaces/contexts?

aghaisas replied:
No, it doesn't. I have updated JDK-8263363 to include this correction.

mrserb commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLPaints.m:
```diff
@@ -0,0 +1,993 @@
+{
+    initTemplatePipelineDescriptors();
+    // This block is not reached in current implementation.
+    // Texture paint XOR mode rendering uses a tile based rendering using a SW pipe (similar to OGL)
```
Do we have a bugid to implement this later?

mrserb commented on src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLSurfaceDataBase.h:
```diff
@@ -0,0 +1,150 @@
+ *     jint x/yOffset
+ * The offset in pixels of the Metal viewport origin from the lower-left
+ * corner of the heavyweight drawable.  For example, a top-level frame on
+ * Windows XP has lower-left insets of (4,4).  The Metal viewport origin
```
Do we really use these fields? "Windows XP"??

aghaisas replied:
This will be fixed as a follow-on issue - JDK-8263486.

jayathirthrao commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLSurfaceData.java:
```diff
@@ -0,0 +1,642 @@
+     * and remove the above restrictions, but that would require significantly
+     * more code just to support a few uncommon cases.
+     */
+    public boolean canRenderLCDText(SunGraphics2D sg2d) {
```
Yes Sergey it works fine in 10.14+ systems via metal. Most of the JCK manual tests use LCD text on UI Components and it is recently verified in 10.15+ systems for EA10.

mrserb replied:
Ok, for some reason I thought that the new macOS stopped providing LCD glyphs.

prsadhuk commented on src/java.desktop/macosx/native/libawt_lwawt/awt/AWTSurfaceLayers.m:
```diff
@@ -67,10 +68,11 @@ - (void) setLayer:(CALayer *)newLayer {
-// including all OpenGL sublayers
+// Updates back buffer size of the layer if it's an OpenGL/Metal layer
+// including all OpenGL/Metal sublayers
 + (void) repaintLayersRecursively:(CALayer*)aLayer {
```
JDK-8263485

mrserb commented:
> By default we don?t do LCD anti-aliasing for text, only when we set Text Rendering Hint to use LCD antialiasing we take sub-pixel rendering path.

And it actually produces LCD glyphs? not the grayscale?

prsadhuk commented on src/java.desktop/macosx/classes/sun/awt/CGraphicsConfig.java:
```diff
@@ -80,7 +80,7 @@ public AffineTransform getNormalizingTransform() {
      * Creates a new SurfaceData that will be associated with the given
      * CGLLayer.
      */
```
I guess we need to specify CGLLayer/MTLLayer as now we are implementing this method in both pipeline.

aghaisas replied:
I have updated JDK-8263363 to include this.

aghaisas commented:
/integrate

avu commented on src/java.desktop/macosx/classes/sun/java2d/metal/MTLBlitLoops.java:
```diff
@@ -0,0 +1,905 @@
+            cachedSrc = srcTmp.get();
+        }
+
+        // We can convert argb_pre data from MTL surface in two places:
```
If we're talking about metal blit operation, the answer is No. Also, it looks like we still need a blit operation from video memory even if we use compute shader.
